### ISSUE TITLE
Deprecate special case behavior for non-None IProtocol.dataReceived return value handling

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2491]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2491|
|Type|task|
|Created|2007-02-27 17:36:23Z|
|Branch|https://github.com/twisted/twisted/tree/no-datareceived-result-2491|

There is no documentation for what IProtocol.dataReceived may return.  Most of the time, ```None``` is the only valid return value.  However, some transports handle other return values to mean other special things (like a reason for disconnection).  We should just get rid of this behavior, after a reasonable period of deprecation.

Attachments:

* [2491.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82f/82f752807893c63f06db6a25fe56160403214fb3/f8986159bccc8c03edfda1a17f3ff7918dd1bb2c.patch) (3477 bytes) - added by michaelnt on 2011-01-30 22:20:16Z - 
* [2491-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82f/82f752807893c63f06db6a25fe56160403214fb3/6b5539c0fef55112d2983ba7553739ca660498e3.patch) (4140 bytes) - added by michaelnt on 2011-02-15 19:19:37Z - 
* [deprecateFunction.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82f/82f752807893c63f06db6a25fe56160403214fb3/2e2d1d9d3389275cfdbfaaf1734a90be2e3adcca.patch) (7458 bytes) - added by michaelnt on 2011-03-05 11:58:41Z - 
* [registry.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82f/82f752807893c63f06db6a25fe56160403214fb3/ec3a042227876d203a59ec449a127ce106f7b4ac.patch) (9106 bytes) - added by michaelnt on 2011-03-16 08:01:07Z - 
* [no-module-globals.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82f/82f752807893c63f06db6a25fe56160403214fb3/2508e5408db60022e79ca37bc24002ec3ddc661b.patch) (6513 bytes) - added by michaelnt on 2011-03-16 08:34:09Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2491 2491
type__task task
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_no_datareceived_result_2491 branches/no-datareceived-result-2491
branch_author__michaelnt michaelnt
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1172597783000000 1172597783000000
changetime__1300317153000000 1300317153000000
version__None None
owner__ 
cc__binjured
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

The ticket summary says "deprecate and remove" but let's just say this ticket is only for the "deprecate" phase.

This would be nice to take care of.  It's one of the ugly corners of `IProtocol` that will hamper adoption of the interface by other people (something I would like to start pushing).

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set owner to @djfroofy<br>@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="binjured's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|binjured commented|
|-|-|

I'd like to take this over since it sounds both easy and important, but there are a lot of places that dataReceived() is called. Generally, I can only find one place in twisted.internet that it is actually returned: t.i.tcp[#7192](https://github.com/twisted/twisted/issues/7192) in Connection.doRead(). Is this ticket asking that the return value be checked and a deprecation warning given in the case that it is not None? Or am I completely off the mark?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> Is this ticket asking that the return value be checked and a deprecation warning given in the case that it is not None?

Yes, just that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to binjured<br>@glyph set status to `new`|
|-|-|

Looks like you're the winner, then.

**Comment by twisted-trac** [comment]
|<img alt="binjured's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|binjured commented|
|-|-|

This ticket is currently held up by me not knowing where to start whitebox testing for the tcp module. Literally, there is the matter of where the module should go; exarkun has recommended _t.i.tcp.tests.test_tcp_connection_. Further, I am not sure how one would test _t.i.tcp.Connection.doRead_ as I have insufficient experience testing low-level, internal APIs.

Exarkun's feedback in IRC:

  * Probably just instantiate Connection()
  * Use a fake socket; there is no current implementation and this should be "verified" (?)

Any guidance or a place to look for similar implementations would be appreciated. I'm familiar with emitting / checking for warnings, just not the specifics of test scaffolding for low-level classes.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|michaelnt removed owner|
|-|-|

Added a patch which issues a DeprecationWarning if self.protocol.dataReceived returns something other than None and tests of this.

Not sure if the version string in the warning is the right version to use.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

Replying to [michaelnt](#note_8):
> Not sure if the version string in the warning is the right version to use.

That should be 11.0, also see [ReleaseProcess#Versionnumbers].

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|michaelnt commented|
|-|-|

Updated the Twisted version number in the patch to 11,0,0.

P.S. Ran the complete Twisted test suite and this warning does not get raised by any of the existing tests.

**Comment by twisted-trac** [comment]
|<img alt="binjured's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|binjured commented|
|-|-|

L[#7434](https://github.com/twisted/twisted/issues/7434), the assert still assumes a deprecation version of 10.3. I didn't notice anything else, but then I didn't know how to write the TestCase in the first place ;)

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|michaelnt commented|
|-|-|

Oops update the patch so that the Twisted version in the test is also 11,0,0

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to michaelnt|
|-|-|

Hello again michaelnt, thanks for your many contributions :).

  1. The stuff in `twisted.internet.test.test_tcpconnection` should mostly go in `twisted.internet.test.test_tcp`.  However, there should be a division between "fake stuff just for one test" (like the `IProtocol` that returns a non-`None` value) and "generalized in-memory implementations of other interfaces (like the "fake" socket).  The latter category of stuff should probably live in `twisted.test.proto_helpers`, which is an unfortunately named module, but it's where all of our other in-memory stand-ins for things like the reactor and transports go, so it's the natural place for such a socket.
  2. All the tests need docstrings, explaining what they're testing and why.  For example, there are lots of other tests which test `doRead`, so `test_doread` should say which case exactly it's covering and what the expected behavior is.  There should definitely never be any empty or blank docstrings within Twisted.

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|michaelnt removed owner|
|-|-|

1.OK I moved the tests to twisted.internet.test.test_tcp. Is it worth trying to use the FakeSocket outside of this test? Wouldn't it need to implement more of the socket API to be generally useful. If I implement a more complete mock would it be used anywhere?


 2. Added some doc strings and renamed the test method.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [30850]) Branching to 'no-datareceived-result-2491'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [30851]) Apply 2491-1.patch

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [30860]) Minor reformatting to stay within 80 columns

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to michaelnt|
|-|-|

Thanks michael.

I think we should have a module for `FakeSocket` and such things.  I don't think it's a pressing issue to get it in the right place immediately.  This `FakeSocket` doesn't look like it will do much that will be generally useful quite yet.

I reformatted some code to get it into <80 columns, and tweaked some docstrings.  More importantly though, I think emitting the deprecation warning with `warnings.warn` is going to give confusing results:

```
/home/exarkun/Projects/Twisted/trunk/twisted/internet/selectreactor.py:146: DeprecationWarning: Returning a value other than None from __main__.P.dataReceived was deprecated in Twisted 11.0.0
  why = getattr(selectable, method)()
```

All this really has going for it is that `__main__.P.dataReceived` is in there.  That's enough to get people to the right place, but the location leading up to that and the line following it are just lies.

If we use `warn_explicit` instead we can point the warning at the definition of `__main__.P.dataReceived`:

```
foo.py:7: DeprecationWarning: Returning a value other than None from __main__.P.dataReceived is deprecated since Twisted 11.0.0.
  def dataReceived(self, bytes):
```

This is a little trickier, since `warn_explicit` requires more arguments than `warn`, and some of them take some effort to get.  To prove to myself that it was really possible to do this with `warn_explicit` I implemented it, committed in r30863.  I didn't actually implement one critical piece though, the part that finds the right line number to pass to `warn_explicit`.  It also probably needs more tests, because maybe the `module` and `registry` and `module_globals` arguments can also affect the behavior (like what gets suppressed, whether duplicate warnings are emitted, etc).

Unless anyone disagrees about this form of the warning being more desirable, I think we should add the necessary `getfirstlineno` function to make this work and write a few more tests for those other args being correct.

Thanks again for working on this michael!

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|michaelnt removed owner|
|-|-|

Factored exarkun's code into a new deprecate.deprecateFunction and added tests.

Modified the code to work with python 2.4, tested on python 2.4 and 2.6.

Perhaps this code should live somewhere else?

There is some duplication between this new code and the TestCase.flushWarnings.

**Comment by twisted-trac** [comment]
|<img alt="waterbug's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|waterbug commented|
|-|-|

Since this is a patch by a non-committer, it shouldn't show up in the easy category, because other non-committer reviewers can't commit it.  Committers:  feel free to put the "easy" keyword back after you review it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

This is quite related to [#2215](https://github.com/twisted/twisted/issues/2215).

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_24"></a>|@djfroofy set owner to michaelnt|
|-|-|

I would like to review this but I think the patch needs to be redone - doesn't apply cleanly against a clean checkout of trunk now:

```
patching file twisted/internet/tcp.py
Reversed (or previously applied) patch detected!  Assume -R? [n]
```

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort set owner to @washort<br>@washort set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|michaelnt commented|
|-|-|

The patch is against the branch not against trunk

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_27"></a>|@washort set owner to michaelnt<br>@washort set status to `new`|
|-|-|

Looks pretty good; there's a few things that should change, though:

1. `FakeSocket` has a `blocking` attribute; real sockets don't, so testing for its value seems unproductive. 
2. `deprecateFunction` sounds like a name for a decorator for deprecation. Maybe a better name would be `blame`?
3. `deprecateFunction`'s docstring says it takes a callable, but it's only tested on functions (and probably only makes sense for functions).

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_28"></a>|@washort commented|
|-|-|

don't mind me, i'm just gaming the highscore system

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_30"></a>|michaelnt removed owner|
|-|-|

Replying to [washort](#note_27):
> Looks pretty good; there's a few things that should change, though:
> 
> 1. `FakeSocket` has a `blocking` attribute; real sockets don't, so testing for its value seems unproductive. 

t.i.tcp.Connection requires a socket to have a setblocking method so it is required.

> 2. `deprecateFunction` sounds like a name for a decorator for deprecation. Maybe a better name would be `blame`?

blame is pretty vague, and only DeprecationWarnings are issued. deprecateFunction seems to fit better with the existing deprecateAttribute, but feel free to change it whilst committing.

> 3. `deprecateFunction`'s docstring says it takes a callable, but it's only tested on functions (and probably only makes sense for functions). 

I don't think that's worth a patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

(In [31094]) Apply deprecateFunction.patch

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun commented|
|-|-|

(In [31098]) whitespace-cleanup

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

(In [31100]) Some whitespace changes, some docstring twiddling

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [31101]) Comment out some critical parts of the implementation which are not covered by the unit tests

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [31102]) Adjust the usage of this function to match the renaming

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun set owner to michaelnt<br>@exarkun set status to `new`|
|-|-|

Thanks michael!

I did some of the boring changes, whitespace and tweaked some naming and docstrings.  I think the only thing that's really left to do is add testing for the last three parameters to `warnings.warn_explicit`.  I checked in a revision that comments them out, and the current tests all still pass.

It's _possible_ that the values I was passing aren't even really necessary, I'm not totally sure.  I think `__warningregistry__` provides the feature of suppressing the second and later instances of a particular warning.  I'm not sure about `__module__` and `module_globals` though.  If we can't figure out what difference they make, I guess we shouldn't bother trying to pass any particular value for this parameters.

**Comment by twisted-trac** [comment]
|<img alt="michaelnt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_38"></a>|michaelnt removed owner|
|-|-|

Please ignore the registry patch, unless you're really interested in trying to load lines of source code from modules loaded from zipfiles and displaying these lines in warning messages (which unittest doesn't support).

The no-module-globals.patch adds tests for the registry attribute and sets module_globals to None, which is what the C (builtin) implemetation of warnings.warn does.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

(In [31185]) Apply no-module-globals.patch 

refs [#2491](https://github.com/twisted/twisted/issues/2491)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun commented|
|-|-|

Thanks!  I am sad about the behaviour of the C implementation of warnings.warn but I guess I am happy it means we don't have to do extra work to implement more silly features. ;)

There was one issue with `sys.modules` state being shared between some tests in `twisted.trial` and `twisted.python` and causes failures.  I fixed it by resetting `sys.modules` completely at the end of the test and that seemed to help.

I'm going to test this on buildbot and if that goes well, merge.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

Argh, I should have known, Windows filename issues.  I'll see if they're easy to fix.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun commented|
|-|-|

Across a number of revisions far greater than should have been necessary, the tests work on Windows now.  It was just necessary to construct path strings using the native directory separator (I let `FilePath` do it) and then `normcase` before comparing.  This was a test change only.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_43"></a>|@exarkun set status to `closed`|
|-|-|

(In [31210]) Merge no-datareceived-result-2491

Author: michaelnt, exarkun
Reviewer: exarkun, washort
Fixes: [#2491](https://github.com/twisted/twisted/issues/2491)

Deprecate the disconnection behavior of returning any non-None value
from `IProtocol.dataReceived`.

Also add a warning helper function, `warnAboutFunction`, for emitting
a warning "at" a function which is not currently on the call stack.
